File: examples/rest-quickstart-nodejs/call_rest_api.js
#!/usr/bin/env node

/**
 * Make one authenticated REST API call to the BDP endpoint.
 *
 * This script is a minimal transport check and first data read for consumers.
 * It supports three resources:
 *
 * - organizations
 * - sites
 * - buildings (requires --site-id)
 *
 * Usage:
 *     node call_rest_api.js --resource sites --take 5
 *     node call_rest_api.js --resource organizations
 *     node call_rest_api.js --resource buildings --site-id YOUR_SITE_GUID
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { parseArgs } from 'util';
import https from 'https';
import { URL } from 'url';
import dotenv from 'dotenv';

const TOKEN_ENV = 'BDP_API_TOKEN';
const UAT_BASE_URL = 'https://ecostruxure-building-platform-api-uat.se.app';
const RESOURCE_TO_PATH = {
  organizations: '/api/Organizations',
  sites: '/api/Sites',
  buildings: '/api/Buildings',
};

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Read token from a dotenv file, or null when absent/unreadable.
 */
function tokenFromDotenv(dotenvPath) {
  try {
    const content = fs.readFileSync(dotenvPath, 'utf-8');
    for (const rawLine of content.split('\n')) {
      const line = rawLine.trim();
      if (!line || line.startsWith('#') || !line.includes('=')) {
        continue;
      }
      const [name, ...valueParts] = line.split('=');
      if (name.trim() !== TOKEN_ENV) {
        continue;
      }
      let token = valueParts.join('=').trim();
      if (
        (token.startsWith('"') && token.endsWith('"')) ||
        (token.startsWith("'") && token.endsWith("'"))
      ) {
        token = token.slice(1, -1);
      }
      return token || null;
    }
  } catch (err) {
    // File not readable or doesn't exist
    return null;
  }
  return null;
}

/**
 * Read token from env first, then fallback to .env in local/script directory.
 */
function resolveToken() {
  const token = process.env[TOKEN_ENV];
  if (token) {
    return token;
  }

  const scriptDirEnv = path.join(__dirname, '.env');
  const cwdEnv = path.join(process.cwd(), '.env');

  let resolvedToken = tokenFromDotenv(scriptDirEnv);
  if (resolvedToken) {
    return resolvedToken;
  }

  // Support running from a directory different from this script's folder.
  if (path.resolve(scriptDirEnv) !== path.resolve(cwdEnv)) {
    return tokenFromDotenv(cwdEnv);
  }
  return null;
}

/**
 * Build the full URL with query parameters.
 */
function buildUrl(baseUrl, resource, siteId, take, skip) {
  const resourcePath = RESOURCE_TO_PATH[resource];
  if (!resourcePath) {
    throw new Error(`Unknown resource: ${resource}`);
  }

  if (resource === 'buildings' && !siteId) {
    throw new Error('--site-id is required when --resource buildings is used');
  }

  const query = new URLSearchParams();
  query.append('take', String(take));
  query.append('skip', String(skip));

  if (resource === 'buildings') {
    query.append('siteId', siteId);
  }

  const url = new URL(baseUrl);
  url.pathname = resourcePath;
  url.search = query.toString();
  return url.toString();
}

/**
 * Print a compact, human-first summary of known response shapes.
 */
function summarizeJson(payload) {
  if (Array.isArray(payload)) {
    console.log(`items: ${payload.length}`);
    for (const item of payload.slice(0, 10)) {
      if (typeof item === 'object' && item !== null) {
        const itemId = item.id ?? '-';
        const itemName = item.name ?? '-';
        console.log(`  - ${itemId} | ${itemName}`);
      } else {
        console.log(`  - ${item}`);
      }
    }
    return;
  }

  if (typeof payload === 'object' && payload !== null) {
    if (Array.isArray(payload.items)) {
      console.log(`items: ${payload.items.length}`);
      for (const item of payload.items.slice(0, 10)) {
        if (typeof item === 'object' && item !== null) {
          const itemId = item.id ?? '-';
          const itemName = item.name ?? '-';
          console.log(`  - ${itemId} | ${itemName}`);
        } else {
          console.log(`  - ${item}`);
        }
      }
      return;
    }

    const fields = Object.keys(payload).sort().join(', ');
    console.log(`top-level fields: ${fields}`);
    return;
  }

  console.log(typeof payload);
}

/**
 * Make an HTTPS request and return the response status and parsed JSON body.
 */
async function requestJson(url, token, apiVersion, timeout) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);

    const options = {
      hostname: urlObj.hostname,
      port: urlObj.port || 443,
      path: urlObj.pathname + urlObj.search,
      method: 'GET',
      headers: {
        Authorization: `Bearer ${token}`,
        'X-Api-Version': apiVersion,
        Accept: 'application/json',
      },
      timeout,
    };

    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        let payload;
        try {
          payload = data ? JSON.parse(data) : {};
        } catch (err) {
          payload = { raw: data };
        }
        resolve({ status: res.statusCode, payload });
      });
    });

    req.on('error', (err) => {
      reject(err);
    });

    req.on('timeout', () => {
      req.destroy();
      reject(new Error('Request timeout'));
    });

    req.end();
  });
}

/**
 * Print error message based on HTTP status code and body.
 */
function printError(code, body) {
  const snippet = body.substring(0, 400).trim().replace(/\n/g, ' ');

  if (code === 401) {
    console.error(
      'HTTP 401: token expired or malformed. Copy a fresh token from the portal.'
    );
    return;
  }
  if (code === 403 && body.includes('Application-Gateway')) {
    console.error(
      'HTTP 403: refused before the API. The request did not reach the API.'
    );
    return;
  }
  if (code === 400 && body.includes('Unsupported API Version')) {
    console.error('HTTP 400: unsupported API version. Use 2.0 or 3.0.');
    return;
  }

  console.error(`HTTP ${code}: ${snippet}`);
}

/**
 * Parse command-line arguments.
 */
async function main(argv) {
  const options = {
    resource: {
      type: 'string',
      default: 'sites',
      short: 'r',
    },
    'site-id': {
      type: 'string',
      short: 's',
    },
    take: {
      type: 'string',
      default: '5',
    },
    skip: {
      type: 'string',
      default: '0',
    },
    'base-url': {
      type: 'string',
      default: UAT_BASE_URL,
    },
    'api-version': {
      type: 'string',
      default: '3.0',
    },
    timeout: {
      type: 'string',
      default: '30',
    },
    raw: {
      type: 'boolean',
      default: false,
    },
    help: {
      type: 'boolean',
      short: 'h',
    },
  };

  let parsed;
  try {
    parsed = parseArgs({
      args: argv,
      options,
      allowPositionals: false,
      strict: true,
    });
  } catch (err) {
    console.error(err.message);
    console.error(
      '\nUsage: node call_rest_api.js [--resource sites|organizations|buildings] [--take 5] [--skip 0] [--api-version 3.0] [--timeout 30] [--raw]'
    );
    process.exit(2);
  }

  if (parsed.values.help) {
    console.log('Usage: node call_rest_api.js [OPTIONS]');
    console.log('\nOptions:');
    console.log('  --resource organizations|sites|buildings   (default: sites)');
    console.log('  --site-id <GUID>                           required for buildings');
    console.log('  --take <int>                               (default: 5)');
    console.log('  --skip <int>                               (default: 0)');
    console.log('  --base-url <url>                           (default: UAT host)');
    console.log('  --api-version 2.0|3.0                      (default: 3.0)');
    console.log('  --timeout <seconds>                        (default: 30)');
    console.log('  --raw                                      print full JSON response');
    console.log('  --help                                     show this help message');
    process.exit(0);
  }

  const resource = parsed.values.resource.toLowerCase();
  const siteId = parsed.values['site-id'];
  const take = parseInt(parsed.values.take, 10);
  const skip = parseInt(parsed.values.skip, 10);
  const baseUrl = parsed.values['base-url'];
  const apiVersion = parsed.values['api-version'];
  const timeoutSeconds = parseInt(parsed.values.timeout, 10);
  const raw = parsed.values.raw;

  // Validate resource choice
  if (!RESOURCE_TO_PATH[resource]) {
    console.error(
      `--resource must be one of: ${Object.keys(RESOURCE_TO_PATH).join(', ')}`
    );
    process.exit(2);
  }

  // Validate numeric arguments
  if (isNaN(take) || take < 0 || isNaN(skip) || skip < 0) {
    console.error('--take and --skip must be non-negative integers');
    process.exit(2);
  }

  if (isNaN(timeoutSeconds) || timeoutSeconds <= 0) {
    console.error('--timeout must be a positive integer');
    process.exit(2);
  }

  // Validate API version
  if (apiVersion !== '2.0' && apiVersion !== '3.0') {
    console.error('--api-version must be 2.0 or 3.0');
    process.exit(2);
  }

  // Load .env file if it exists
  dotenv.config({ path: path.join(__dirname, '.env') });
  dotenv.config({ path: path.join(process.cwd(), '.env') });

  // Resolve token
  const token = resolveToken();
  if (!token) {
    console.error(
      `set ${TOKEN_ENV} in your environment or add it to a .env file in this folder`
    );
    process.exit(2);
  }

  // Build URL
  let url;
  try {
    url = buildUrl(baseUrl, resource, siteId, take, skip);
  } catch (err) {
    console.error(err.message);
    process.exit(2);
  }

  console.log(`GET ${url}`);

  // Make the request
  try {
    const { status, payload } = await requestJson(
      url,
      token,
      apiVersion,
      timeoutSeconds * 1000
    );

    if (status < 200 || status >= 300) {
      const body = typeof payload === 'string' ? payload : JSON.stringify(payload);
      printError(status, body);
      process.exit(1);
    }

    console.log(`status: ${status}`);
    if (raw) {
      console.log(JSON.stringify(payload, null, 2));
    } else {
      summarizeJson(payload);
    }

    process.exit(0);
  } catch (err) {
    if (err.message === 'Request timeout') {
      console.error('network error: request timed out');
    } else if (err.code === 'ENOTFOUND') {
      console.error(`network error: ${err.message}`);
    } else {
      console.error(`network error: ${err.message}`);
    }
    process.exit(1);
  }
}

// Run the script
main(process.argv.slice(2));

File: examples/rest-quickstart-python/call_rest_api.py
"""Make one authenticated REST API call to the BDP endpoint.

This script is a minimal transport check and first data read for consumers.
It supports three resources:

- organizations
- sites
- buildings (requires --site-id)

Usage:
    python call_rest_api.py --resource sites --take 5
    python call_rest_api.py --resource organizations
    python call_rest_api.py --resource buildings --site-id YOUR_SITE_GUID
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

TOKEN_ENV = "BDP_API_TOKEN"
UAT_BASE_URL = "https://ecostruxure-building-platform-api-uat.se.app"
RESOURCE_TO_PATH = {
    "organizations": "/api/Organizations",
    "sites": "/api/Sites",
    "buildings": "/api/Buildings",
}


def token_from_dotenv(dotenv_path: Path) -> str | None:
    """Return token from a dotenv file, or None when absent/unreadable."""
    try:
        with dotenv_path.open(encoding="utf-8") as handle:
            for raw_line in handle:
                line = raw_line.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                name, value = line.split("=", 1)
                if name.strip() != TOKEN_ENV:
                    continue
                token = value.strip()
                if token.startswith(("\"", "'")) and token.endswith(("\"", "'")):
                    token = token[1:-1]
                return token or None
    except OSError:
        return None
    return None


def resolve_token() -> str | None:
    """Read token from env first, then fallback to .env in local/script directory."""
    token = os.environ.get(TOKEN_ENV)
    if token:
        return token

    script_dir_env = Path(__file__).with_name(".env")
    cwd_env = Path.cwd() / ".env"

    token = token_from_dotenv(script_dir_env)
    if token:
        return token

    # Support running from a directory different from this script's folder.
    if script_dir_env.resolve() != cwd_env.resolve():
        return token_from_dotenv(cwd_env)
    return None


def build_url(base_url: str, resource: str, site_id: str | None,
              take: int, skip: int) -> str:
    path = RESOURCE_TO_PATH[resource]
    query: dict[str, str] = {"take": str(take), "skip": str(skip)}

    if resource == "buildings":
        if not site_id:
            raise ValueError("--site-id is required when --resource buildings is used")
        query["siteId"] = site_id

    encoded = urllib.parse.urlencode(query)
    return f"{base_url.rstrip('/')}{path}?{encoded}"


def summarize_json(payload: object) -> None:
    """Print a compact, human-first summary of known response shapes."""
    if isinstance(payload, list):
        print(f"items: {len(payload)}")
        for item in payload[:10]:
            if isinstance(item, dict):
                item_id = item.get("id", "-")
                item_name = item.get("name", "-")
                print(f"  - {item_id} | {item_name}")
            else:
                print(f"  - {item}")
        return

    if isinstance(payload, dict):
        if "items" in payload and isinstance(payload["items"], list):
            items = payload["items"]
            print(f"items: {len(items)}")
            for item in items[:10]:
                if isinstance(item, dict):
                    item_id = item.get("id", "-")
                    item_name = item.get("name", "-")
                    print(f"  - {item_id} | {item_name}")
                else:
                    print(f"  - {item}")
            return

        print("top-level fields:", ", ".join(sorted(payload.keys())))
        return

    print(type(payload).__name__)


def request_json(url: str, token: str, api_version: str,
                 timeout: int) -> tuple[int, object]:
    request = urllib.request.Request(
        url,
        headers={
            "Authorization": f"Bearer {token}",
            "X-Api-Version": api_version,
            "Accept": "application/json",
        },
        method="GET",
    )

    with urllib.request.urlopen(request, timeout=timeout) as response:
        status = response.status
        body = response.read().decode("utf-8", errors="replace")
        try:
            payload = json.loads(body) if body else {}
        except json.JSONDecodeError:
            payload = {"raw": body}
    return status, payload


def print_error(code: int, body: str) -> None:
    snippet = body[:400].strip().replace("\n", " ")
    if code == 401:
        print(
            "HTTP 401: token expired or malformed. Copy a fresh token from the portal.",
            file=sys.stderr,
        )
        return
    if code == 403 and "Application-Gateway" in body:
        print(
            "HTTP 403: refused before the API. The request did not reach the API.",
            file=sys.stderr,
        )
        return
    if code == 400 and "Unsupported API Version" in body:
        print(
            "HTTP 400: unsupported API version. Use 2.0 or 3.0.",
            file=sys.stderr,
        )
        return

    print(f"HTTP {code}: {snippet}", file=sys.stderr)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--resource",
        choices=sorted(RESOURCE_TO_PATH),
        default="sites",
        help="resource to request",
    )
    parser.add_argument("--site-id", help="required when --resource buildings")
    parser.add_argument("--take", type=int, default=5)
    parser.add_argument("--skip", type=int, default=0)
    parser.add_argument(
        "--base-url",
        default=UAT_BASE_URL,
        help="REST API host, no /api suffix needed",
    )
    parser.add_argument(
        "--api-version",
        default="3.0",
        choices=("2.0", "3.0"),
    )
    parser.add_argument("--timeout", type=int, default=30)
    parser.add_argument(
        "--raw",
        action="store_true",
        help="print full JSON response instead of a compact summary",
    )
    args = parser.parse_args(argv)

    if args.take < 0 or args.skip < 0:
        print("--take and --skip must be non-negative", file=sys.stderr)
        return 2

    token = resolve_token()
    if not token:
        print(
            f"set {TOKEN_ENV} in your environment or add it to a .env file in this folder",
            file=sys.stderr,
        )
        return 2

    try:
        url = build_url(args.base_url, args.resource, args.site_id, args.take, args.skip)
    except ValueError as exc:
        print(str(exc), file=sys.stderr)
        return 2

    print(f"GET {url}")

    try:
        status, payload = request_json(url, token, args.api_version, args.timeout)
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", errors="replace")
        print_error(exc.code, body)
        return 1
    except urllib.error.URLError as exc:
        print(f"network error: {exc.reason}", file=sys.stderr)
        return 1

    print(f"status: {status}")
    if args.raw:
        print(json.dumps(payload, indent=2, ensure_ascii=False))
    else:
        summarize_json(payload)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
